presto-cli/src/main/java/com/facebook/presto/cli/StatusPrinter.java

```diff
@@ -443,4 +457,92 @@ private static double percentage(double count, double total)
         }
         return min(100, (count * 100.0) / total);
     }
+
+    private class ProgressStats
+    {
+        private final long timestampNanos;
+        private final long processedBytes;
+        private final long processedRows;
+
+        public ProgressStats(long timestampNanos, long processedBytes, long processedRows)
+        {
+            this.timestampNanos = timestampNanos;
+            this.processedBytes = processedBytes;
+            this.processedRows = processedRows;
+        }
+
+        public long getTimestampNanos()
+        {
+            return timestampNanos;
+        }
+
+        public long getProcessedBytes()
+        {
+            return processedBytes;
+        }
+
+        public long getProcessedRows()
+        {
+            return processedRows;
+        }
+    }
+
+    private class ProgressStatsWindow
+    {
+        private final long windowSize;
+        private final LinkedList<ProgressStats> progressWindow = new LinkedList<>();
+
+        public ProgressStatsWindow(long windowSize)
+        {
+            this.windowSize = windowSize;
+        }
+
+        public void addProgressStats(ProgressStats ps)
+        {
+            progressWindow.add(ps);
+            while (progressWindow.getFirst().getTimestampNanos() < ps.getTimestampNanos() - windowSize) {
+                progressWindow.removeFirst();
+            }
+        }
+
+        public String getDataRate()
+        {
+            Duration elapsedTime = succinctNanos(progressWindow.getLast().getTimestampNanos() - progressWindow.getFirst().getTimestampNanos());
+            return formatDataRate(bytes(progressWindow.getLast().getProcessedBytes() - progressWindow.getFirst().getProcessedBytes()), elapsedTime, false);
+        }
+
+        public String getRowCountRate()
+        {
+            Duration elapsedTime = succinctNanos(progressWindow.getLast().getTimestampNanos() - progressWindow.getFirst().getTimestampNanos());
+            return formatCountRate(progressWindow.getLast().getProcessedRows() - progressWindow.getFirst().getProcessedRows(), elapsedTime, false);
+        }
+    }
+
+    private class StatementProgressStats
```

Comment:
I don't think the class name should be `StatementProgressStats`, it contains both stage progress stats 

Maybe `ProgressStatsHistory` ?